File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return [[]]
        res = []
        for i in range(0, len(nums)):
            for sub in self.permute(nums[0:i:] + nums[i+1::]):
                res.append([nums[i]] + sub)
        return res

File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    if (nums.length === 1) return [nums]
    const res = []
    for (let i=0; i < nums.length; i++) {
        for (const sub of permute(nums.slice(0, i).concat(nums.slice(i+1)))) {
            res.push([nums[i], ...sub])
        }
    }
    return res
};